New module to test UI API against metab example flowsheet (#527)
watertap-org/watertap: docs/how_to_guides/how_to_use_ui_api.rst

.. _howto_ui-api:

How-to guide for the UI API
===========================
.. py:currentmodule:: watertap.ui.api

.. contents:: Contents
    :depth: 2
    :local:

Overview
--------

There are two distinct intended users for this API:

.. image:: /_static/terminal-icon.png
    :height: 30px
    :align: left

Model developers  can select which variables to "export" to the UI for each component of the model, and provide extra metadata (display name, description) for them.
For flowsheets, they also should specify how to build and solve the flowsheets.

.. image:: /_static/menu-icon.png
    :height: 22px
    :align: left

User interface developers can "discover" flowsheets for inclusion in the UI, and use the API to provide JSON versions of their contents (and results).

The rest of this page will provide more detail on how to do tasks for each type of user.

See also: :ref:`the UI API reference section <ref_ui-api>`.

Model Developers
----------------

.. image:: /_static/terminal-icon.png
    :height: 30px
    :align: left

Exporting variables with :func:`export_variables`
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> For each IDAES component (also a Pyomo block) the developer should list the names (and, optionally some additional information) of the variables that should be "exported" to the user interface.

Reviewer: ```suggestion
For each IDAES component and/or Pyomo block the developer should list the names of the variables that should be "exported" to the user interface. 
Optionally, users may provide additional information as well (see :func:`export_variables` for details).
```

This should be done in the component's ``build`` method.
Each component/block need only worry about its own variables.

> Teh API will take care of gathering all the components for a flowsheet together.

Reviewer: ```suggestion
The API will take care of gathering all the components for a flowsheet together.
```

For example, the last line of the zero-order feed's `build` method is::

    export_variables(self, name="Feed Z0", desc="Zero-Order feed block",
                         variables=["flow_vol", "conc_mass_comp"])

At the time of that call, ``self`` is the feed block that was just built.
It is saying to export two variables called "flow_vol" and "conc_mass_comp".

There are other things about variables that can be specified.
For details see :func:`export_variables`.

> For example, if you wanted to mark "flow_vol" as read-only (loading it into the model will not change it), then you could do::

Reviewer: ```suggestion
For example, if you wanted to mark "flow_vol" as read-only, then you could do::
```

    export_variables(self, name="Feed Z0", desc="Zero-Order feed block",
                         variables={"flow_vol": {"readonly": True},
                                    "conc_mass_comp": {}})

> 

Reviewer: ```suggestion

**Note:** Marking a variable as read-only does change the IDAES variable.

```

3. :ref:`Set the action functions into the flowsheet interface <howto_api-set-action>` before returning it to the user.

.. image:: /_static/terminal-icon.png
    :height: 30px
    :align: left

.. _howto_api-ui_define-interface:

Define the ``flowsheet_interface`` function
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
The flowsheet interface is created by a function with this signature:

.. function:: flowsheet_interface() -> FlowsheetInterface

   In other words, it takes no arguments and returns a :class:`FlowsheetInterface` object.
   This object is not yet connected to an IDAES flowsheet block.
   The function should return a flowsheet interface, see :ref:`howto_api-ui_create-interface`.

Note that you only need to add variables that are not already exported by the model, and that there are reasonable defaults for things like the name, display_name (same as name), and description. So in most cases this will be a very simple call; the extended version was shown for didactic purposes.

.. image:: /_static/terminal-icon.png
    :height: 30px
    :align: left

.. _howto_api-ui_create-action:

Create an "action" function
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> The action functions should operate on optional keywords for the flowsheet block and the FlowsheetInterface instance.

Reviewer: ```suggestion
The action functions should operate on optional keywords for the flowsheet block and the :class:`FlowsheetInterface` instance.
```

It will call the WaterTAP code to perform the appropriate actions.
The function should have the following signature:

.. function:: action_function([block=None, ui=None], **kwargs)

    Perform an action.

    :param Block block: Flowsheet block
    :param FlowsheetInterface ui: FlowsheetInterface instance
    :param dict kwargs: Additional key/value pairs specific to this action

For example::

>     def build_flowsheet(ui=None, **kwargs):

Reviewer: This example snippet may need some additional explanation. 

Where did `my_model` come in from (or is this just meant to be a more generic representation of any model actually being built in this function)? 

It also did not become clear to me how this `build_flowsheet` function is actually used until the last instruction. Maybe consider adding a link to the next instruction saying that this function will be used within your `flowsheet_interface` function.

Author: Right, I see how this is confusing. `my_model` was supposed to represent a module (like the metab module) that has functions to build and run the full treatment train. I modified the code to explicitly add an import, and renamed it `my_flowsheet`:

```
    # import module for the full flowsheet
    from watertap.examples.flowsheets.case_studies.foo.bar import my_flowsheet
```

I also added the link, I agree that helps.

By default this function will find all interfaces in the ``watertap`` Python package in which it is situated::

    from watertap.ui.api import find_flowsheet_interfaces
    for fsi in find_flowsheet_interfaces():
        print(f"Got flowsheet: {fsi.display_name}")

You can add a configuration file or dict to specify alternate or additional places to look.
Just remember when doing so to provide the default watertap location if you want it included.
For example::

    from watertap.ui.api import find_flowsheet_interfaces
    interface_list = find_flowsheet_interfaces(config={
        "packages": ["watertap", "my_other_package"]})
    for fsi in interface_list:
        print(f"Got flowsheet: {fsi.display_name}")

.. image:: /_static/menu-icon.png
    :height: 22px
    :align: left

.. _howto_api-fetch-update:

Fetch and update flowsheet values
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
The values for all variables exported by the flowsheet are available via the ``dict()`` method.
The format of the returned value is documented in the :mod:`~watertap.ui.api` module header.

> Note: if you want to write these values as JSON to a stream, use :meth:`FlowsheetInterface.save`.

Reviewer: ```suggestion

**Note:** if you want to write these values as JSON to a stream, use :meth:`FlowsheetInterface.save`.
```

    for fsi in find_flowsheet_interfaces():
        print(f"Got flowsheet: {fsi.display_name}")
        print(f"Flowsheet contents: {fsi.dict()}")

.. image:: /_static/menu-icon.png
    :height: 22px
    :align: left

.. _howto_api-run-actions:

Run flowsheet actions
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Once flowsheet actions are created, you invoke them with the :meth:`FlowsheetInterface.run_action` method. The name of the action, if it is a standard one, will be an attribute in the :class:`WorkflowActions` class. For example::

    from watertap.ui.api import FlowsheetInterface

    def run_build(fsi: FlowsheetInterface):
        fsi.run_action(WorkflowActions.build)

.. image:: /_static/menu-icon.png
    :height: 22px
    :align: left

.. _howto_api-save-flowsheet:

Save flowsheet values in a file
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
The values of variables that were :func:`exported <export_variables>` for a given flowsheet can be saved to a file.
For example::

    from watertap.ui.api import FlowsheetInterface
    fsi = FlowsheetInterface()
    fsi.set_block(my_flowsheet)
    fsi.save("my_flowsheet_saved.json")

> Note: The method is simply a wrapper that calls ``dict()`` and feeds the result to a JSON serializer.

Reviewer: ```suggestion
**Note:** The method is simply a wrapper that calls ``dict()`` and feeds the result to a JSON serializer.
```

.. image:: /_static/menu-icon.png
    :height: 22px
    :align: left

.. _howto_api-load-flowsheet:

Load flowsheet values from a file
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
The values of variables that were :ref:`saved <howto_api-save-flowsheet>` for a given flowsheet can be loaded back into the model.
This operation changes the matching values in the actual underlying model to the values that are stored in the file.
In other words, unlike the :meth:`FlowsheetInterface.save` method, this method changes what is stored in memory.
Invoking load is straightforward::

    from watertap.ui.api import FlowsheetInterface
    # Create and save
    fsi = FlowsheetInterface()
    fsi.set_block(my_flowsheet)
    fsi.save("my_flowsheet_saved.json")
    # Load back
    fsi.load("my_flowsheet_saved.json")

> To handle situations where the model changes over time, and the saved data and model do not match exactly, there are two methods you can use after you are done loading:

Reviewer: Maybe we need to qualify what is meant by '... model changes over time' to avoid confusion for dynamic models (which are functions of time). 

Do you mean that this is to handle a model that has changed due to some user update or code update? Or do you actually mean a model that is time-dependent?

Author: I meant the former.  Time-dependent models are not something I have tested with, and that is something that should be looked at, since I assumed that indexes are discrete (for starters). But for this section, I have modified the wording to:
```
To handle situations where the model is changed between the save and subsequent load, so that the loaded and current model variables do not all match, there are two methods you can use after you are done loading:

    :meth:`FlowsheetInterface.get_var_missing`
        Returns variables in the loaded data, but not in the current model.

    :meth:`FlowsheetInterface.get_var_extra`
        Returns variables in (and exported by) the current model, but not in the loaded data.

Both methods return a mapping with the full name of the block (e.g., `flowsheet.component.subcomponent`) as the key and a list of variable names as the value.
```
(Apologies for the long lines, I am letting my editor soft-wrap the lines for me and only inserting line breaks for new sentences -- avoids manual reformatting and makes diffs cleaner.)
